src/test/java/com/google/jenkins/plugins/computeengine/integration/SpotVmProvisioningWithMaxRunDurationCasCIT.java

```diff
@@ -0,0 +1,162 @@
+
+        // verify no nodes to start with.
+        assertEquals(0, j.jenkins.getNodes().size());
+
+        ComputeEngineCloud cloud = (ComputeEngineCloud) j.jenkins.clouds.getByName("gce-integration");
+        cloud.getConfigurations().get(0).setGoogleLabels(vmLabels);
+        Collection<PlannedNode> planned = cloud.provision(new LabelAtom(LABEL), 1);
+        planned.iterator().next().future.get(); // wait for the node creation to finish
+        Instance instance =
+                client.getInstance(PROJECT_ID, ZONE, planned.iterator().next().displayName);
+        String instanceName = instance.getName();
+        log.info("Instance: " + instance.getName());
+        assertEquals(1, j.jenkins.getNodes().size());
+
+        // assert the scheduling configurations.
+        Scheduling sch = instance.getScheduling();
+        assertEquals("SPOT", sch.get("provisioningModel"));
+        assertEquals("DELETE", sch.get("instanceTerminationAction"));
+        assertEquals(180, Integer.parseInt((String) ((ArrayMap) sch.get("maxRunDuration")).get("seconds")));
+        log.info("instance scheduling configs are correct");
+
+        // try to execute a build on the agent (freestyle type)
+        FreeStyleProject fp = j.createFreeStyleProject();
+        Builder step = execute(Commands.ECHO, "hello world");
+        fp.getBuildersList().add(step);
+        fp.setAssignedLabel(new LabelAtom(LABEL));
+        Future<FreeStyleBuild> buildFuture = fp.scheduleBuild2(0);
+        FreeStyleBuild build = buildFuture.get();
+        assertEquals(Result.SUCCESS, build.getResult());
+        String agent1 = printLogsAndReturnAgentName(build.getLog(50));
+        log.info("first build completed on" + agent1);
+        assertEquals(agent1, instanceName);
+
+        // wait for the agent to be removed due to maxRunDuration; retention time is far more 30 minutes - so the
+        // deletion should be due to maxRunDuration.
+        await("Jenkins agent to be removed due to maxRunDuration well before the retention time")
+                .atMost(180, TimeUnit.SECONDS)
+                .until(() -> j.jenkins.getNodes().isEmpty());
+        await("GCP VM deletion to complete due to maxRunDuration")
+                .atMost(180, TimeUnit.SECONDS)
+                .until(() -> client.listInstancesWithLabel(PROJECT_ID, vmLabels).isEmpty());
+
+        // trigger another build, notice a new instance is being created (pipeline type)
+        log.info("proceeding to 2nd build, after no remaining instances");
+        var p = j.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("node('" + LABEL + "') { sh 'date' }", true));
+        var run = j.assertBuildStatus(Result.SUCCESS, p.scheduleBuild2(0).get());
+        String agent2 = printLogsAndReturnAgentName(run.getLog(50));
+        log.info("second build completed on " + agent2);
+        assertNotEquals(agent1, agent2);
+        log.info("spot vm test completed");
+    }
```

Comment:
There is a warning from `RemainingActivityListener` in the quoted log output. If still there now, make sure you are waiting for all builds to complete.

Reply:
This is fixed now, added `j.waitUntilNoActivity()`